qiskit/circuit/library/data_preparation/state_preparation.py

```diff
@@ -414,3 +439,143 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
             circuit.append(CXGate(), [msb, lsb])
 
         return circuit
+
+
+class Generalized_Uniform_Superposition_Gate(Gate):
+    """
+    Class that implements a generalized uniform superposition state, using n qubits, following the Shukla-Vedula algorithm [SV24].
+    """
+
+    def __init__(
+        self,
+        M : int = 2,
+        num_qubits: Optional[int] = None,
+    ):
+        """
+        Initializes Generalized_Uniform_Superposition_Gate.
+    
+        Args:
+            M (int): 
+                A positive integer M (> 1) representing the number of computational basis states with an amplitude of 1/sqrt(M) 
+                in the uniform superposition state ($\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $, where $1< M <= 2^n$). Note that the remaining 
+                (2^n - M) computational basis states have zero amplitudes. Here M need not be an integer power of 2.
+                
+            num_qubits (int): 
+                A positive integer representing the number of qubits used.
+        
+        Returns:
+            A quantum circuit that creates the uniform superposition state: $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $. 
+    
+        **References:**
+            [SV24] 
+                A. Shukla and P. Vedula, "An efficient quantum algorithm for preparation of uniform quantum superposition states,"
+                Quantum Information Processing, 23(38): pp. 1-32 (2024).
+
+        Raises:
+            QiskitError: ``num_qubits`` parameter used when ``params`` is not an integer
+                
+        Note:
+            The Shukla-Vedula algorithm [SV24] provides an efficient approach for creation of a generalized uniform superposition 
+            state of the form, $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $. It needs only $O(\log_2 (M))$ qubits and $O(\log_2 (M))$ 
+            gates, hence providing an exponential improvement, in terms of reduced resources and complexity, compared to alternative methods
+            in existing literature.
+        """
+        if not(isinstance(M, int) and (M > 1)):
+            raise ValueError(
+                "M must be a positive integer greater than 1."
+            )
+        if num_qubits is None:
+            num_qubits = np.ceil(np.log2(M))
```

Comment:
Make sure you add a unit test that has the num_qubits as None so it tests this. I am not sure whether this returns an int type exactly - I know some instance checks we had for Python int etc would fail - in places I have seen the return from ceil cast to int so I am not sure. It may fail in the super class when this is passed to it in the super init call.

Reply:
Oh wow, there is a case when `M` is a power of 2 and that would not be handled well. Thanks for pointing it out, I'll include the test case when `num_qubits = None` .